hooks/src/index.js

```diff
@@ -164,18 +164,35 @@ export function useReducer(reducer, initialState, init) {
 			!init ? invokeOrReturn(undefined, initialState) : init(initialState),
 
 			action => {
-				const nextValue = hookState._reducer(hookState._value[0], action);
-				if (hookState._value[0] !== nextValue) {
-					hookState._value = [nextValue, hookState._value[1]];
+				const currentValue = hookState._nextValue
+					? hookState._nextValue[0]
+					: hookState._value[0];
+				const nextValue = hookState._reducer(currentValue, action);
+
+				if (currentValue !== nextValue) {
+					hookState._nextValue = [
+						hookState._reducer(currentValue, action),
+						hookState._value[1]
+					];
 					hookState._component.setState({});
 				}
 			}
 		];
 
 		hookState._component = currentComponent;
+
+		hookState._component.shouldComponentUpdate = () => {
```

Comment:
This does not fix all the cases we still need to do `_nextState` setting for repeating renders https://github.com/preactjs/preact/blob/master/src/diff/index.js#L188